File: src/joints/rapier_joint_2d.cpp
#include "rapier_joint_2d.h"
#include "../spaces/rapier_space_2d.h"

RapierJoint2D::RapierJoint2D(RapierBody2D *p_body_a, RapierBody2D *p_body_b) {
	A = p_body_a;
	B = p_body_b;
}

void RapierJoint2D::copy_settings_from(RapierJoint2D *p_joint) {
	set_rid(p_joint->get_rid());
	set_max_force(p_joint->get_max_force());
	set_bias(p_joint->get_bias());
	set_max_bias(p_joint->get_max_bias());
	disable_collisions_between_bodies(p_joint->is_disabled_collisions_between_bodies());
}

void RapierJoint2D::disable_collisions_between_bodies(const bool p_disabled) {
	disabled_collisions_between_bodies = p_disabled;
	if (rapier2d::is_handle_valid(handle)) {
		// Joint not yet created, when it will be created it will have disable collision flag set
		rapier2d::joint_change_disable_collision(space_handle, handle, is_disabled_collisions_between_bodies());
	}
}

RapierJoint2D::~RapierJoint2D() {
	if (rapier2d::is_handle_valid(handle)) {
		ERR_FAIL_COND(!rapier2d::is_handle_valid(space_handle));
		rapier2d::joint_destroy(space_handle, handle);
	}
};

File: src/joints/rapier_joint_2d.rs
use std::any::Any;

use crate::rapier2d::{
    handle::{invalid_handle, Handle},
    joint::{joint_change_disable_collision, joint_destroy},
};
use godot::{builtin::Rid, engine::physics_server_2d};

use super::{rapier_damped_spring_joint_2d::RapierDampedSpringJoint2D, rapier_pin_joint_2d::RapierPinJoint2D};

pub trait IRapierJoint2D: Any {
    fn get_base(&self) -> &RapierJointBase2D;
    fn get_mut_base(&mut self) -> &mut RapierJointBase2D;
    fn get_type(&self) -> physics_server_2d::JointType;
    fn get_damped_spring(&self) -> Option<&RapierDampedSpringJoint2D>;
    fn get_pin(&self) -> Option<&RapierPinJoint2D>;
    fn get_mut_damped_spring(&mut self) -> Option<&mut RapierDampedSpringJoint2D>;
    fn get_mut_pin(&mut self) -> Option<&mut RapierPinJoint2D>;
}

pub struct RapierJointBase2D {
    max_force: f32,
    rid: Rid,
    handle: Handle,
    space_handle: Handle,
    disabled_collisions_between_bodies: bool,
}

impl RapierJointBase2D {
    pub fn new(space_handle: Handle, handle: Handle, rid: Rid) -> Self {
        Self {
            max_force: f32::MAX,
            rid,
            handle,
            space_handle,
            disabled_collisions_between_bodies: true,
        }
    }
    pub fn get_handle(&self) -> Handle {
        self.handle
    }
    pub fn get_space_handle(&self) -> Handle {
        self.space_handle
    }
    pub fn set_max_force(&mut self, force: f32) {
        self.max_force = force;
    }
    pub fn get_max_force(&self) -> f32 {
        self.max_force
    }
    pub fn get_rid(&self) -> Rid {
        self.rid
    }
    // Careful when doing this you must also update the place where it's stored.
    pub fn set_rid(&mut self, rid: Rid) {
        self.rid = rid;
    }
    pub fn disable_collisions_between_bodies(&mut self, disabled: bool) {
        self.disabled_collisions_between_bodies = disabled;
        if self.handle.is_valid() {
            joint_change_disable_collision(
                self.space_handle,
                self.handle,
                self.disabled_collisions_between_bodies,
            );
        }
    }
    pub fn is_disabled_collisions_between_bodies(&self) -> bool {
        self.disabled_collisions_between_bodies
    }
    pub fn copy_settings_from(&mut self, joint: &RapierJointBase2D) {
        self.set_rid(joint.get_rid());
        self.set_max_force(joint.get_max_force());
        self.disable_collisions_between_bodies(joint.is_disabled_collisions_between_bodies());
    }
}

impl Drop for RapierJointBase2D {
    fn drop(&mut self) {
        if self.handle.is_valid() {
            joint_destroy(self.space_handle, self.handle);
        }
    }
}

pub struct RapierEmptyJoint2D {
    base: RapierJointBase2D,
}

impl RapierEmptyJoint2D {
    pub fn new(rid: Rid) -> Self {
        Self {
            base: RapierJointBase2D::new(invalid_handle(), invalid_handle(), rid),
        }
    }
}

impl IRapierJoint2D for RapierEmptyJoint2D {
    fn get_type(&self) -> physics_server_2d::JointType {
        physics_server_2d::JointType::MAX
    }
    fn get_base(&self) -> &RapierJointBase2D {
        &self.base
    }
    fn get_mut_base(&mut self) -> &mut RapierJointBase2D {
        &mut self.base
    }
    fn get_damped_spring(&self) -> Option<&RapierDampedSpringJoint2D> {
        None
    }
    fn get_pin(&self) -> Option<&RapierPinJoint2D> {
        None
    }
    fn get_mut_damped_spring(&mut self) -> Option<&mut RapierDampedSpringJoint2D> {
        None
    }
    fn get_mut_pin(&mut self) -> Option<&mut RapierPinJoint2D> {
        None
    }
}
